src/platform/backends/lxd/lxd_request.cpp

```diff
@@ -34,6 +34,13 @@ namespace
 {
 constexpr auto request_category = "lxd request";
 
+void log_and_throw_error(const std::string& message, mpl::Level level)
+{
+    mpl::log(level, request_category, message);
+
+    throw std::runtime_error(message);
+}
```

Comment:
This feels weird to be here in `lxd_request.cpp` only. Shouldn't this be a general thing?

Or really, shouldn't whatever catches the `runtime_error` log it, if it doesn't bubble it up? I know you'd lose control over the logging level, but if needed, that could be done with different exception types?

Reply:
Hmm, yeah, I was trying to combine redundant code, but yeah, after reading your feedback, it should up to the caller to log anything as it sees fit.  The callee here should just throw the error.

Reply:
Ok, I've been working on this and all I end up doing is instead of having one place that logs a message for our purposes, I'm spreading the same thing over and over to different parts of the code, making it redundant.  If you don't like `log_and_throw_error()` here, I can move it to `utils.cpp`, but I think this function commonizes something we need.

Reply:
Ok, I removed this function and made a generic function in `utils.h`.  If we want to log file and line numbers and not get the function in utils, then the message that is passed in should contain the file and line and that is what will be logged.

Reply:
I removed this function and also got rid of the templated function introduced in the previous now defunct commit.  We also discussed using a large `try`/`catch` encapsulating `lxd_request()` and `lxd_wait()` with the `catch` logging the message and then rethrowing.  However, I found that if we want logging level control, a new exception type will need to be introduced to accept the logging level.  When I did this, it became clear that the exception itself should just do the logging which is what I have now.

Reply:
Admittedly I didn't follow this discussion or read the code closely, so apologies if something here is not applicable :slightly_smiling_face: but I am intrigued. Normally it should be up to some caller up in the stack to decide how to handle an exception, including whether it can be recovered from, whether it should be logged, with what severity, etc. 

Exceptions are a way for the code where the problem occurs to abort what it's doing but delegate handling to callers, providing any relevant specifics in the exception itself, but letting "someone above", where all the context is available, judge how to react. As far as the thrower knows, proper handling could be switching backends, trying a different server, showing a GUI popup, reporting to a remote entity, retrying... 

Assuming an exception is an error at the throw site disables code reuse and approaches of "asking for forgiveness rather than permission". So logging in the exception constructor sounds strange to me, especially if it's logged with more severity then `debug`. [Jon Kalb](http://www.exceptionsafecode.com/) is a good source of material about this, his talks on error handling are elucidative.

Reply:
Hey @ricab,

Thanks for the feedback!  It is definitely still applicable:smile:

So here's the problem I'm *trying* to solve.  In `lxd_request_common()` and `lxd_wait()`, there are a number of places where an "issue" can occur in the request of an API endpoint in the LXD daemon as can be seen in all the places that `throw`.  Right now, all but one "issue" I want to log at the `error` severity, so I deemed it necessary for the code to need to set the severity level for logging when throwing the exception.  Of course, there are a few ways to handle this.  My approach is to try to keep code redundancy to a minimum which is why I've gone the way of using common `log_and_throw`-type functions or the `LXDRuntimeError` which just encapsulates the loggig in itself.

That said, since this awkward, I could instead keep the `LXDRuntimeError` exception type, and any caller of functions that throw this type can just assume it's an error and choose to log.  Really, there is only one place where I set `warning` instead of `error` and I could argue the `warning` should be an `error` as well.

@saviq suggested a `try`/`catch`/(re)`throw` around the functions much like we do with the daemon command functions like `launch`.  I think I might try that approach coupled with just making the callers log any caught `LXDRuntimeError`'s at the `error` level.

Reply:
OK, I think I understand your point. Looking from afar, that seems like a sensible option (single `error` severity from a handler up in the stack). If that meant giving up too much, you could introduce separate exception types. But if you find that the simpler code is worth giving up the distinction, better still. 

Anyway, I just wanted to provide some feedback, but I'm leaving this to you guys.

Reply:
Ok, after some more thinking and discussion with @ricab, I've decided to just use the `try`/`catch`/(re)`throw` methodology to log messages.

Reply:
Marking this as resolved since I'm using the `try`/`catch` method.